File: Backtracking_and_Search/LC_131_Palindrome_Partitioning.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        ans = []
        
        def is_palid(s):
            return s == s[::-1]
        def dfs(s, curr, ans):
            #print(s)
            if len(s) == 0:
                ans.append(curr[:])
                return 
            for i in range(len(s)):
                first, second = s[:i+1], s[i+1:]
                if is_palid(first):
                    curr.append(first)
                    dfs(second, curr, ans)
                    curr.pop()
        dfs(s, [], ans)
        return ans

File: Backtracking_and_Search/LC_131_Palindrome_Partitioning.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> curr;
        dfs(s, curr, ans);
        return ans;
    }
private:
    bool is_palid(string s){
        int l = 0, r = s.size() - 1;
        while(l < r){
            if (s[l] != s[r]){
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
    void dfs(string s, vector<string>& curr, vector<vector<string>>& ans){
        if (s.size() == 0){
            ans.push_back(curr);
            return;
        }
        for (int l = 1; l <= s.size(); l++){
            string first = s.substr(0, l), second = s.substr(l);
            //cout << first <<" " << second << endl;
            if (is_palid(first)){
                curr.push_back(first);
                dfs(second, curr, ans);
                curr.pop_back();
            }
        }


    }
};